extension/observer/dockerobserver/extension.go

```diff
@@ -91,21 +93,71 @@ func (d *dockerObserver) ListAndWatch(listener observer.Notify) {
 	d.emitContainerEndpoints(listener)
 
 	go func() {
+		// build filter list for container events
+		filters := dfilters.NewArgs([]dfilters.KeyValuePair{
+			{Key: "type", Value: "container"},
+			{Key: "event", Value: "destroy"},
+			{Key: "event", Value: "die"},
+			{Key: "event", Value: "pause"},
+			{Key: "event", Value: "rename"},
+			{Key: "event", Value: "stop"},
+			{Key: "event", Value: "start"},
+			{Key: "event", Value: "unpause"},
+			{Key: "event", Value: "update"},
+		}...)
+
+		lastTime := time.Now()
 		ticker := time.NewTicker(d.config.CacheSyncInterval)
 		defer ticker.Stop()
+
+	EVENT_LOOP:
 		for {
-			select {
-			case <-d.ctx.Done():
-				return
-			case <-ticker.C:
-				err := d.syncContainerList(listener)
-				if err != nil {
-					d.logger.Error("Could not sync container cache", zap.Error(err))
+			opts := dtypes.EventsOptions{
+				Filters: filters,
+				Since:   lastTime.Format(time.RFC3339Nano),
+			}
+
+			eventsCh, errCh := d.dClient.Events(d.ctx, opts)
+
+			for {
+				select {
+				case <-d.ctx.Done():
+					return
+				case <-ticker.C:
+					err := d.syncContainerList(listener)
+					if err != nil {
+						d.logger.Error("Could not sync container cache", zap.Error(err))
+					}
+				case event := <-eventsCh:
+					switch event.Action {
+					case "destroy":
+						d.updateEndpointsByContainerID(listener, event.ID, nil)
+						d.dClient.RemoveContainer(event.ID)
+					default:
+						if c, ok := d.dClient.InspectAndPersistContainer(d.ctx, event.ID); ok {
+							endpointsMap := d.endpointsForContainer(c)
+							d.updateEndpointsByContainerID(listener, c.ID, endpointsMap)
+						}
+					}
+					if event.TimeNano > lastTime.UnixNano() {
+						lastTime = time.Unix(0, event.TimeNano)
+					}
+
+				case err := <-errCh:
+					// if there's an error from docker events we need to
+					// reinvoke the Events function to begin streaming again.
+					if d.ctx.Err() == nil {
```

Comment:
Is this correct condition? not `!= nil`?

Reply:
This loop is based on the existing client's loop - this err == nill check is for 
> We are only interested in re-connecting when the context hasn't been cancelled, since requests made with a closed context are guaranteed to fail.